05-mvc-with-rails/error-handling/readme.md

```diff
@@ -3,14 +3,461 @@
+
+### Else
+
+What we really want is for the controller to do one thing when it works, and another when it doesn't.
+
+`begin/rescue` blocks can also take an `else` statement. We can use this to, say, show the Artist show page if there are no errors, but redirect the user to Google if there are still errors:
+
+```rb
+  def create
+    begin
+      @artist = Artist.create!(artist_params)
+    rescue StandardError => sad_panda
+      redirect_to "http://google.com/search?q=#{sad_panda.message}"
+    else
+      redirect_to "/artists/#{@artist.id}"
+    end
+  end
+```
+
+### Raising your own errors
+
+Underneath the surface, errors all rely on the `raise` keyword. You don't "throw" an error in Ruby, you "raise" it.
+
+For example:
+
+```rb
+  def create
+    begin
+      @artist = Artist.create!(artist_params)
+      if @artist.name == "Billy Ray Cyrus"
+        raise "An Artist who isn't Billy Ray Cyrus, please"
+      end
+    rescue StandardError => sad_panda
+      redirect_to "http://google.com/search?q=#{sad_panda.message}"
+    else
+      redirect_to "/artists/#{@artist.id}"
+    end
+  end
+```
+
+Whenever Ruby encounters a `raise`, it stops whatever it's doing and looks to see whether the `raise` occurred inside a `begin/rescue` block. If it did, then it lets the `rescue` handle the error. If it did **not**, then it "breaks" the app with a fatal error.
+
+You can raise an error anywhere you want. It will always work the same way.
+
+# Flash
+
+## Framing
+
+...but we don't *want* to redirect the user to Google. We just want to show the error message. So let's do that the way we did before:
+
+```rb
+  def create
+    begin
+      @artist = Artist.create!(artist_params)
+    rescue StandardError => sad_panda
+      @error = sad_panda.message
+    else
+      redirect_to "/artists/#{@artist.id}"
+    end
+  end
+```
+
+We'll get a "missing template" error if we don't give the controller something to render when it has errors, so let's have it show the `new` view again:
+
+```rb
+  def create
+    begin
+      @artist = Artist.create!(artist_params)
+    rescue StandardError => sad_panda
+      @error = sad_panda.message
+      render :new
+    else
+      redirect_to "/artists/#{@artist.id}"
+    end
+  end
+```
+
+That's pretty good. But I'm not a huge fan of the "Confirm Form Submission" box that pops up when you try to refresh a page that was rendered via POST.
+
+Instead, we can have Rails just redirect the user to whichever page they were on before:
+
+```rb
+  def create
+    begin
+      @artist = Artist.create!(artist_params)
+    rescue StandardError => sad_panda
+      @error = sad_panda.message
+      redirect_to :back
+    else
+      redirect_to "/artists/#{@artist.id}"
+    end
+  end
```

Comment:
Working on a PR to head us in that direction.  Maybe begin/rescue/end is example outside of rails.  I'm having trouble coming up with "real-life" example.  Maybe when connecting to 3rd party?